File: src/main/java/io/github/Chase22/telegram/telegrambotapimock/infrastructure/server/endpoint/EndpointRegistry.java
package io.github.Chase22.telegram.telegrambotapimock.infrastructure.server.endpoint;

import io.github.Chase22.telegram.telegrambotapimock.api.endpoints.GetMeEndpoint;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

public class EndpointRegistry {
    private final Logger LOGGER = LoggerFactory.getLogger(EndpointRegistry.class);

    private Map<String, Endpoint> endpointMap = new HashMap<>();

    public EndpointRegistry() {
        registerEndpoints();
        registerTelegramEndpoints();
    }

    public Endpoint getForPath(String path) {
        return endpointMap.get(path);
    }

    private void registerEndpoints() {
        addEndpoint(new HealthEndpoint());
    }

    private void registerTelegramEndpoints() {
        addEndpoint(new GetMeEndpoint());
    }

    private void addEndpoint(Endpoint servlet) {
        LOGGER.info("Registering servlet " + servlet.getClass().getName() + " with mapping " + servlet.getPath());
        endpointMap.put(servlet.getPath(), servlet);
    }
}

File: src/main/kotlin/io/github/chase22/telegram/telegrambotapimock/infrastructure/registry/EndpointRegistry.kt
package io.github.chase22.telegram.telegrambotapimock.infrastructure.registry

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.chase22.telegram.telegrambotapimock.telegram.data.Update
import io.github.chase22.telegram.telegrambotapimock.telegram.data.User
import io.github.chase22.telegram.telegrambotapimock.telegram.endpoints.DeleteWebhook
import io.github.chase22.telegram.telegrambotapimock.telegram.endpoints.GetMe
import io.github.chase22.telegram.telegrambotapimock.telegram.endpoints.GetUpdates
import io.github.chase22.telegram.telegrambotapimock.telegram.endpoints.TelegramApiEndpoint
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.BlockingQueue

class EndpointRegistry(objectMapper: ObjectMapper, bot: User, updateChannel: BlockingQueue<Update>) {

    private val endpointMap = HashMap<String, TelegramApiEndpoint<*>>()

    init {
        addEndpoint(GetMe(bot, objectMapper))
        addEndpoint(DeleteWebhook(objectMapper))
        addEndpoint(GetUpdates(updateChannel, objectMapper))
    }

    fun getForPath(path: String): TelegramApiEndpoint<*>? = endpointMap[path]

    private fun addEndpoint(servlet: TelegramApiEndpoint<*>) {
        LOGGER.info("Registering servlet " + servlet.javaClass.name + " with mapping " + servlet.path)
        endpointMap[servlet.path] = servlet
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(EndpointRegistry::class.java)
    }
}
